### PR TITLE
Addresses bad behavior with overridden optimizer.step by #20124

### PR DESCRIPTION
**Pull request by vfdev-5**
This PR addresses the problem described in the comment: https://github.com/pytorch/pytorch/pull/20203#issuecomment-499231276
and previously coded bad behaviour:
- a warning was raised all the times when lr schedulling is initialized

Now the code checks that:
- on the second call of `lr_scheduler.step`, ensure that `optimizer.step` has been already called, otherwise raise a warning (as it was done in #20203 )
- if optimizer's step is overridden -> raise once another warning to aware user about the new pattern:
`opt.step()` -> `lrs.step()` as we can not check this .

Now tests check that 
- at initialization (`lrs = StepLR(...)`)there is no warnings
- if we replace `optimizer.step` by something else (similarly to the [code of nvidia/apex](https://github.com/NVIDIA/apex/blob/master/apex/amp/_process_optimizer.py#L287)) there is another warning raised.

cc @ezyang 

PS. honestly I would say that there is a lot of overhead introduced for simple warnings. I hope all these checks will be removed in future `1.2.0` or other versions...

### CODE REVIEW

**Comment by ezyang** [comment]
> honestly I would say that there is a lot of overhead introduced for simple warnings. I hope all these checks will be removed in future 1.2.0 or other versions...

Yeah, I think that would be the best idea.

**Comment by ezyang** [comment]
Or, if you think we should just can the warnings immediately, because they are too much work to implement (and aren't going to help our most affected users, those on 1.1 release), I'd also be amenable to that too.

**Comment by ezyang** [comment]
I mentioned this to @soumith and he said we definitely should not remove the warning :)

File: test/test_optim.py [hunk 1]
```diff
@@ -603,6 +646,31 @@ def test_new_pattern_no_warning_with_arg(self):
                 scheduler.step(e)
             self.assertTrue(len(ws) == 0, "No warning should be raised")
 
+    def test_new_pattern_no_warning_with_overriden_optim_step(self):
```
Comment (ezyang) [inline test/test_optim.py]: @mcarilli Could you confirm that this test case accurately captures the wrapping that APEX is doing on step?

**Review by ezyang — APPROVED** [review]

File: test/test_optim.py (hunk 1 above)
Comment (mcarilli) [reply test/test_optim.py]: Looks reasonable to me.  For iterations that Amp chooses to skip, Amp patches optimizer.step to a new method that does not carry out an update, but instead, restores optimizer.step to the original method.

File: torch/optim/lr_scheduler.py [hunk 2]
```diff
@@ -63,13 +65,23 @@ def get_lr(self):
...
         # Raise a warning if old pattern is detected
         # https://github.com/pytorch/pytorch/issues/20124
-        if self.optimizer.step.called < 1:
-            warnings.warn("Detected call of `lr_scheduler.step()` before `optimizer.step()`. "
-                          "In PyTorch 1.1.0 and later, you should call them in the opposite order: "
-                          "`optimizer.step()` before `lr_scheduler.step()`.  Failure to do this "
-                          "will result in PyTorch skipping the first value of the learning rate schedule."
-                          "See more details at "
-                          "https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate", UserWarning)
+        if self._step_count == 1:
```
Comment (stevenwjy) [inline torch/optim/lr_scheduler.py:68]: Hello author @vfdev-5, sorry to trouble you, but just curious what's the reasoning behind checking whether it's equal to 1 here? Shouldn't it be `self._step_count == 0` (checking whether this is the first call to the function or not)? Thanks!

Comment (vfdev-5) [reply torch/optim/lr_scheduler.py:68]: Hello, the first call (`self._step_count == 0`) happens in `__init__` (line 43). See the details in the linked issue: https://github.com/pytorch/pytorch/issues/20124

Comment (stevenwjy) [reply torch/optim/lr_scheduler.py:68]: Ah I see, sorry didn't notice that... Thanks for clarifying!

**Pull request closed without merging**